File: banking_system.py
"""
Banking System Simulation (CodeSignal-style, Levels 1-4)
=========================================================
Time-ordered stream of operations -> one string result per query.

Design (first principles)
-------------------------
State:
  balance[acc]   -> int                current balance
  outgoing[acc]  -> int                lifetime outgoing (TRANSFER + PAY)
  payments[pid]  -> {owner, at, cb}    owner = CURRENT owner (survives merges)
  acc_payments[acc] -> list[pid]       payment ids owned by acc (for merge)
  cashback_heap  -> min-heap of (credit_time, pid)   pending cashbacks

Key idea: lazy settlement. Before EVERY query, pop all heap entries with
credit_time <= timestamp and credit floor(2%) to payments[pid]["owner"].
The heap stores payment ids (not account ids), so MERGE only has to
re-point payment ownership — pending heap entries follow automatically.

Complexity: n = #queries <= 1e4.
  Each op O(log n) amortized (heap), TOP_SPENDERS O(A log A),
  MERGE O(#payments of merged account). Total well within limits.
"""

import heapq
from typing import Dict, List

CASHBACK_DELAY_MS = 86_400_000  # 24h
CASHBACK_PCT = 2


class BankingSystem:
    def __init__(self) -> None:
        self.balance: Dict[str, int] = {}
        self.outgoing: Dict[str, int] = {}
        self.payments: Dict[str, dict] = {}          # pid -> {owner, at, cb}
        self.acc_payments: Dict[str, List[str]] = {} # acc -> [pid]
        self.cashback_heap: List[tuple] = []         # (credit_time, pid)
        self.pay_counter = 0

    # ---------- lazy cashback settlement (run before every op) ----------
    def _settle(self, ts: int) -> None:
        while self.cashback_heap and self.cashback_heap[0][0] <= ts:
            _, pid = heapq.heappop(self.cashback_heap)
            p = self.payments[pid]
            self.balance[p["owner"]] += p["cb"]
            p["done"] = True

    # ---------- Level 1 ----------
    def create_account(self, ts: int, acc: str) -> str:
        if acc in self.balance:
            return "false"
        self.balance[acc] = 0
        self.outgoing[acc] = 0
        self.acc_payments[acc] = []
        return "true"

    def deposit(self, ts: int, acc: str, amount: int) -> str:
        if acc not in self.balance:
            return ""
        self.balance[acc] += amount
        return str(self.balance[acc])

    def get_balance(self, ts: int, acc: str) -> str:
        if acc not in self.balance:
            return ""
        return str(self.balance[acc])

    # ---------- Level 2 ----------
    def transfer(self, ts: int, src: str, dst: str, amount: int) -> str:
        if (src not in self.balance or dst not in self.balance
                or src == dst or self.balance[src] < amount):
            return ""
        self.balance[src] -= amount
        self.balance[dst] += amount
        self.outgoing[src] += amount
        return str(self.balance[src])

    def top_spenders(self, ts: int, n: int) -> str:
        ranked = sorted(self.outgoing.items(), key=lambda kv: (-kv[1], kv[0]))
        return ", ".join(f"{acc}({total})" for acc, total in ranked[:n])

    # ---------- Level 3 ----------
    def pay(self, ts: int, acc: str, amount: int) -> str:
        if acc not in self.balance or self.balance[acc] < amount:
            return ""
        self.balance[acc] -= amount
        self.outgoing[acc] += amount
        self.pay_counter += 1
        pid = f"payment{self.pay_counter}"
        self.payments[pid] = {
            "owner": acc,
            "at": ts,
            "cb": amount * CASHBACK_PCT // 100,
            "done": False,
        }
        self.acc_payments[acc].append(pid)
        heapq.heappush(self.cashback_heap, (ts + CASHBACK_DELAY_MS, pid))
        return pid

    def get_payment_status(self, ts: int, acc: str, pid: str) -> str:
        if acc not in self.balance:
            return ""
        p = self.payments.get(pid)
        if p is None or p["owner"] != acc:
            return ""
        return "CASHBACK_RECEIVED" if p["done"] else "IN_PROGRESS"

    # ---------- Level 4 ----------
    def merge_accounts(self, ts: int, a1: str, a2: str) -> str:
        if a1 == a2 or a1 not in self.balance or a2 not in self.balance:
            return "false"
        self.balance[a1] += self.balance[a2]
        self.outgoing[a1] += self.outgoing[a2]
        # Redirect payment ownership -> pending cashbacks & status follow.
        for pid in self.acc_payments[a2]:
            self.payments[pid]["owner"] = a1
        self.acc_payments[a1].extend(self.acc_payments[a2])
        del self.balance[a2], self.outgoing[a2], self.acc_payments[a2]
        return "true"


def solution(queries: List[List[str]]) -> List[str]:
    bank = BankingSystem()
    ops = {
        "CREATE_ACCOUNT":     lambda ts, a: bank.create_account(ts, a),
        "DEPOSIT":            lambda ts, a, amt: bank.deposit(ts, a, int(amt)),
        "GET_BALANCE":        lambda ts, a: bank.get_balance(ts, a),
        "TRANSFER":           lambda ts, s, d, amt: bank.transfer(ts, s, d, int(amt)),
        "TOP_SPENDERS":       lambda ts, n: bank.top_spenders(ts, int(n)),
        "PAY":                lambda ts, a, amt: bank.pay(ts, a, int(amt)),
        "GET_PAYMENT_STATUS": lambda ts, a, p: bank.get_payment_status(ts, a, p),
        "MERGE_ACCOUNTS":     lambda ts, a1, a2: bank.merge_accounts(ts, a1, a2),
    }
    results = []
    for op, ts_str, *args in queries:
        ts = int(ts_str)
        bank._settle(ts)  # apply all cashbacks due at/before ts
        results.append(ops[op](ts, *args))
    return results


# ---------------------------- tests ----------------------------
if __name__ == "__main__":
    # Provided example
    q = [
        ["CREATE_ACCOUNT", "1", "alice"],
        ["CREATE_ACCOUNT", "2", "bob"],
        ["DEPOSIT", "3", "alice", "2000"],
        ["TRANSFER", "4", "alice", "bob", "500"],
        ["TOP_SPENDERS", "5", "2"],
        ["PAY", "6", "alice", "1000"],
        ["GET_PAYMENT_STATUS", "7", "alice", "payment1"],
        ["GET_PAYMENT_STATUS", "86400007", "alice", "payment1"],
        ["GET_BALANCE", "86400008", "alice"],
    ]
    expected = ["true", "true", "2000", "1500", "alice(500), bob(0)",
                "payment1", "IN_PROGRESS", "CASHBACK_RECEIVED", "520"]
    got = solution(q)
    assert got == expected, f"\ngot:      {got}\nexpected: {expected}"

    # Merge: pending cashback redirected, status queryable via new owner
    q2 = [
        ["CREATE_ACCOUNT", "1", "a"],
        ["CREATE_ACCOUNT", "2", "b"],
        ["DEPOSIT", "3", "b", "1000"],
        ["PAY", "4", "b", "500"],                       # payment1, cb 10 at t=86400004
        ["MERGE_ACCOUNTS", "5", "a", "b"],              # b folded into a
        ["GET_BALANCE", "6", "b"],                      # "" (gone)
        ["GET_PAYMENT_STATUS", "7", "a", "payment1"],   # IN_PROGRESS via a
        ["GET_BALANCE", "86400004", "a"],               # 500 + 10 cashback
        ["GET_PAYMENT_STATUS", "86400005", "a", "payment1"],
        ["TOP_SPENDERS", "86400006", "5"],              # a inherits b's outgoing
    ]
    expected2 = ["true", "true", "1000", "payment1", "true", "",
                 "IN_PROGRESS", "510", "CASHBACK_RECEIVED", "a(500)"]
    got2 = solution(q2)
    assert got2 == expected2, f"\ngot:      {got2}\nexpected: {expected2}"

    # Invalid ops: self-transfer, insufficient funds, missing accounts
    q3 = [
        ["CREATE_ACCOUNT", "1", "x"],
        ["CREATE_ACCOUNT", "2", "x"],           # false (duplicate)
        ["DEPOSIT", "3", "ghost", "100"],       # ""
        ["TRANSFER", "4", "x", "x", "0"],       # "" (self)
        ["TRANSFER", "5", "x", "ghost", "10"],  # "" (missing target)
        ["PAY", "6", "x", "1"],                 # "" (insufficient)
        ["MERGE_ACCOUNTS", "7", "x", "x"],      # false
        ["GET_PAYMENT_STATUS", "8", "x", "payment1"],  # "" (no such payment)
    ]
    expected3 = ["true", "false", "", "", "", "", "false", ""]
    got3 = solution(q3)
    assert got3 == expected3, f"\ngot:      {got3}\nexpected: {expected3}"

    print("All tests passed.")

File: banking_system.cpp
// ============================================================================
// Banking System Simulation (CodeSignal-style, Levels 1-4) — C++17
// ----------------------------------------------------------------------------
// Design (first principles):
//   balance[acc], outgoing[acc]        : core state (int64 — sums can exceed 2^31)
//   payments[pid] = {owner, cb, done}  : owner = CURRENT owner (survives merges)
//   accPayments[acc] = [pid...]        : for O(k) ownership redirect on merge
//   cashbackHeap (min-heap by time)    : pending cashbacks, stores payment ids
//
// Key ideas:
//   1. LAZY SETTLEMENT — before every query, pop all heap entries with
//      creditTime <= ts and credit 2% to payments[pid].owner.
//   2. Heap stores PAYMENT IDS, not accounts — MERGE only re-points
//      payment ownership; pending cashbacks & status follow automatically.
//
// Complexity: n <= 1e4 queries. Each op O(log n) amortized,
//   TOP_SPENDERS O(A log A), MERGE O(#payments of merged account).
// ============================================================================

#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <queue>
#include <algorithm>
#include <utility>
using namespace std;
using ll = long long;

class BankingSystem {
    static constexpr ll CASHBACK_DELAY_MS = 86400000LL; // 24h
    static constexpr ll CASHBACK_PCT = 2;

    struct Payment {
        string owner;   // current owner (updated on merge)
        ll cashback;
        bool done = false;
    };

    unordered_map<string, ll> balance_;
    unordered_map<string, ll> outgoing_;
    unordered_map<string, Payment> payments_;             // pid -> payment
    unordered_map<string, vector<string>> accPayments_;   // acc -> [pid]
    // min-heap of (creditTime, pid)
    priority_queue<pair<ll, string>, vector<pair<ll, string>>, greater<>> cashbackHeap_;
    ll payCounter_ = 0;

public:
    // ------------------------------------------------------------------
    // Lazy cashback settlement — MUST run before every operation.
    // ------------------------------------------------------------------
    void settle(ll ts) {
        while (!cashbackHeap_.empty() && cashbackHeap_.top().first <= ts) {
            auto [t, pid] = cashbackHeap_.top();
            cashbackHeap_.pop();
            Payment& p = payments_[pid];
            balance_[p.owner] += p.cashback;
            p.done = true;
        }
    }

    // ==================================================================
    // Level 1 — Accounts and balances
    // ==================================================================
    string createAccount(ll ts, const string& acc) {
        if (balance_.count(acc)) return "false";
        balance_[acc] = 0;
        outgoing_[acc] = 0;
        accPayments_[acc] = {};
        return "true";
    }

    string deposit(ll ts, const string& acc, ll amount) {
        auto it = balance_.find(acc);
        if (it == balance_.end()) return "";
        it->second += amount;
        return to_string(it->second);
    }

    string getBalance(ll ts, const string& acc) {
        auto it = balance_.find(acc);
        if (it == balance_.end()) return "";
        return to_string(it->second);
    }

    // ==================================================================
    // Level 2 — Transfers and spending ranks
    // ==================================================================
    string transfer(ll ts, const string& src, const string& dst, ll amount) {
        if (src == dst) return "";
        auto s = balance_.find(src), d = balance_.find(dst);
        if (s == balance_.end() || d == balance_.end() || s->second < amount) return "";
        s->second -= amount;
        d->second += amount;
        outgoing_[src] += amount;
        return to_string(s->second);
    }

    string topSpenders(ll ts, ll n) {
        vector<pair<string, ll>> ranked(outgoing_.begin(), outgoing_.end());
        // highest total first; ties -> ascending account id
        sort(ranked.begin(), ranked.end(), [](const auto& a, const auto& b) {
            if (a.second != b.second) return a.second > b.second;
            return a.first < b.first;
        });
        string out;
        for (size_t i = 0; i < ranked.size() && i < (size_t)n; ++i) {
            if (i) out += ", ";
            out += ranked[i].first + "(" + to_string(ranked[i].second) + ")";
        }
        return out;
    }

    // ==================================================================
    // Level 3 — Payments with deferred cashback
    // ==================================================================
    string pay(ll ts, const string& acc, ll amount) {
        auto it = balance_.find(acc);
        if (it == balance_.end() || it->second < amount) return "";
        it->second -= amount;
        outgoing_[acc] += amount;
        string pid = "payment" + to_string(++payCounter_);
        payments_[pid] = {acc, amount * CASHBACK_PCT / 100, false};
        accPayments_[acc].push_back(pid);
        cashbackHeap_.emplace(ts + CASHBACK_DELAY_MS, pid);
        return pid;
    }

    string getPaymentStatus(ll ts, const string& acc, const string& pid) {
        if (!balance_.count(acc)) return "";
        auto it = payments_.find(pid);
        if (it == payments_.end() || it->second.owner != acc) return "";
        return it->second.done ? "CASHBACK_RECEIVED" : "IN_PROGRESS";
    }

    // ==================================================================
    // Level 4 — Account merge
    // ==================================================================
    string mergeAccounts(ll ts, const string& a1, const string& a2) {
        if (a1 == a2 || !balance_.count(a1) || !balance_.count(a2)) return "false";
        balance_[a1] += balance_[a2];
        outgoing_[a1] += outgoing_[a2];
        // Redirect payment ownership: pending cashbacks & status follow.
        auto& moved = accPayments_[a2];
        for (const string& pid : moved) payments_[pid].owner = a1;
        auto& dst = accPayments_[a1];
        dst.insert(dst.end(), moved.begin(), moved.end());
        balance_.erase(a2);
        outgoing_.erase(a2);
        accPayments_.erase(a2);
        return "true";
    }
};

// ============================================================================
// Dispatcher — one result string per query
// ============================================================================
vector<string> solution(const vector<vector<string>>& queries) {
    BankingSystem bank;
    vector<string> results;
    results.reserve(queries.size());

    for (const auto& q : queries) {
        const string& op = q[0];
        ll ts = stoll(q[1]);
        bank.settle(ts); // apply all cashbacks due at/before ts

        if      (op == "CREATE_ACCOUNT")     results.push_back(bank.createAccount(ts, q[2]));
        else if (op == "DEPOSIT")            results.push_back(bank.deposit(ts, q[2], stoll(q[3])));
        else if (op == "GET_BALANCE")        results.push_back(bank.getBalance(ts, q[2]));
        else if (op == "TRANSFER")           results.push_back(bank.transfer(ts, q[2], q[3], stoll(q[4])));
        else if (op == "TOP_SPENDERS")       results.push_back(bank.topSpenders(ts, stoll(q[2])));
        else if (op == "PAY")                results.push_back(bank.pay(ts, q[2], stoll(q[3])));
        else if (op == "GET_PAYMENT_STATUS") results.push_back(bank.getPaymentStatus(ts, q[2], q[3]));
        else if (op == "MERGE_ACCOUNTS")     results.push_back(bank.mergeAccounts(ts, q[2], q[3]));
        else                                 results.push_back("");
    }
    return results;
}

// ============================================================================
// Tests
// ============================================================================
static void check(const vector<vector<string>>& q, const vector<string>& expected,
                  const string& name) {
    vector<string> got = solution(q);
    if (got != expected) {
        cerr << "FAIL: " << name << "\n";
        for (size_t i = 0; i < expected.size(); ++i)
            if (i >= got.size() || got[i] != expected[i])
                cerr << "  [" << i << "] got \"" << (i < got.size() ? got[i] : "<none>")
                     << "\" expected \"" << expected[i] << "\"\n";
        exit(1);
    }
    cout << "PASS: " << name << "\n";
}

int main() {
    // Provided example
    check({
        {"CREATE_ACCOUNT", "1", "alice"},
        {"CREATE_ACCOUNT", "2", "bob"},
        {"DEPOSIT", "3", "alice", "2000"},
        {"TRANSFER", "4", "alice", "bob", "500"},
        {"TOP_SPENDERS", "5", "2"},
        {"PAY", "6", "alice", "1000"},
        {"GET_PAYMENT_STATUS", "7", "alice", "payment1"},
        {"GET_PAYMENT_STATUS", "86400007", "alice", "payment1"},
        {"GET_BALANCE", "86400008", "alice"},
    }, {
        "true", "true", "2000", "1500", "alice(500), bob(0)",
        "payment1", "IN_PROGRESS", "CASHBACK_RECEIVED", "520",
    }, "example");

    // Merge: pending cashback redirected, status queryable via new owner
    check({
        {"CREATE_ACCOUNT", "1", "a"},
        {"CREATE_ACCOUNT", "2", "b"},
        {"DEPOSIT", "3", "b", "1000"},
        {"PAY", "4", "b", "500"},                     // payment1, cb 10 at t=86400004
        {"MERGE_ACCOUNTS", "5", "a", "b"},            // b folded into a
        {"GET_BALANCE", "6", "b"},                    // "" (gone)
        {"GET_PAYMENT_STATUS", "7", "a", "payment1"}, // IN_PROGRESS via a
        {"GET_BALANCE", "86400004", "a"},             // 500 + 10 cashback
        {"GET_PAYMENT_STATUS", "86400005", "a", "payment1"},
        {"TOP_SPENDERS", "86400006", "5"},            // a inherits b's outgoing
    }, {
        "true", "true", "1000", "payment1", "true", "",
        "IN_PROGRESS", "510", "CASHBACK_RECEIVED", "a(500)",
    }, "merge + cashback redirect");

    // Invalid ops
    check({
        {"CREATE_ACCOUNT", "1", "x"},
        {"CREATE_ACCOUNT", "2", "x"},          // false (duplicate)
        {"DEPOSIT", "3", "ghost", "100"},      // ""
        {"TRANSFER", "4", "x", "x", "0"},      // "" (self)
        {"TRANSFER", "5", "x", "ghost", "10"}, // "" (missing target)
        {"PAY", "6", "x", "1"},                // "" (insufficient)
        {"MERGE_ACCOUNTS", "7", "x", "x"},     // false
        {"GET_PAYMENT_STATUS", "8", "x", "payment1"}, // "" (no such payment)
    }, {
        "true", "false", "", "", "", "", "false", "",
    }, "invalid operations");

    cout << "All tests passed.\n";
    return 0;
}
